File: 198_HouseRobber.java
class Solution {
    public int rob(int[] nums) {
        int len = nums.length;
        if (len == 0 || nums == null) return 0;
        if (len == 1) return nums[0];
        int first = nums[0];
        int second = Math.max(first, nums[1]);

        for (int i = 2; i < len; i++) {
            int temp = second;
            second = Math.max(second, first + nums[i]);
            first = temp;
        }
        return second;
    }
}

File: c++/198_HouseRobber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> f(2), g(2);
        for (int i = 1; i <= n; i++) {
            f[i & 1] = g[i - 1 & 1] + nums[i - 1];
            g[i & 1] = max(g[i - 1 & 1], f[i - 1 & 1]);
        }
        return max(g[n & 1], f[n & 1]);
    }
};